### PR TITLE
add: button component and config

### PR DESCRIPTION
**Pull request by MaksymTaratuta55**
https://github.com/ramsayinnovations/ramsay-site/issues/151

### CODE REVIEW

**Comment by MaksymTaratuta55** [comment]
@thecodeclimber, could you please check this PR? Does it what you want?

File: src/components/button/Button.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        shape && theme.button.borderRadius[shape],
+        size && theme.button.size[size],
+        className,
+      ]}
+      {...props}
+    >
+      {children}
+    </button>
+  );
+};
```
Comment (thecodeclimber) [inline src/components/button/Button.js]: It seems to me that we should use the `styled` way of doing this -- i.e., like this:

```
const Button = styled.button(
  ({ block, className, children, color, border, shape, size, ...props }) => [
   theme.button.default.base,
        theme.button.default.color,
        theme.button.default.hover,
        theme.button.default.focus,
        block && tw`w-full h-full`,
        color && theme.button.color[color],
        border === "gray" && theme.button.border[border],
        shape && theme.button.borderRadius[shape],
        size && theme.button.size[size],
        className,
      ]
  )
```

[note, I don't know how we would add in the rest of the props (i.e., `...props`). 

I know this way allows us to use the `as` prop:  https://emotion.sh/docs/styled#as-prop 

With that said, is there any reason to code it the way that you did?  Please let me know.

File: tailwind.config.js [hunk 2]
```diff
@@ -0,0 +1,130 @@
+const defaultTheme = require("tailwindcss/defaultTheme");
+const colors = require("tailwindcss/colors");
+
+module.exports = {
+  future: {
+    purgeLayersByDefault: true,
+  },
+  purge: ["./components/**/*.js", "./pages/**/*.js"],
+  theme: {
```
Comment (thecodeclimber) [inline tailwind.config.js]: I don't think we need a `tailwind.config.js` file in the `Ramsay UI` repository.  We can use the default tailwind configuration.   When we import the library into another project we can then add a `tailwind.config.js` file and pull those values into the `theme.js` file if need be.

**Review by ghost — CHANGES_REQUESTED** [review]
@MaksymTaratuta55 

Please see below some of my questions concerning the code.  

Thanks, Moshe.

File: src/components/button/Button.js (hunk 1 above)
Comment (MaksymTaratuta55) [reply src/components/button/Button.js:30]: updated

File: tailwind.config.js (hunk 2 above)
Comment (MaksymTaratuta55) [reply tailwind.config.js]: Fixed

**Comment by MaksymTaratuta55** [comment]
@thecodeclimber, I've updated with your remarks

File: src/themes/theme.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+      danger: tw`bg-danger-600 hover:bg-danger-700 border-danger-600`,
+    },
+    border: {
+      gray: tw`border border-gray-300 text-gray-700 bg-white-100 hover:bg-gray-50`,
+      grey: tw`border border-gray-300 text-gray-700 bg-white-100 hover:bg-gray-50`,
+      transparent: tw`bg-transparent`,
+      primary: tw`border-primary-500 hover:bg-primary-500 hover:text-gray-50`,
+      success: tw`border-success-500 hover:bg-success-500 hover:text-gray-50`,
+      danger: tw`border-danger-500 hover:bg-danger-500 hover:text-gray-50`,
+    },
```
Comment (thecodeclimber) [inline src/themes/theme.js]: Let's rework the code as follows:

    border: {
      base:  tw`border`,
      gray: tw`border-gray-300 text-gray-700 bg-white-100 hover:bg-gray-50`,
      grey: tw`border-gray-300 text-gray-700 bg-white-100 hover:bg-gray-50`,
      transparent: tw`bg-transparent`,
      primary: tw`border-primary-500 hover:bg-primary-500 hover:text-gray-50`,
      success: tw`border-success-500 hover:bg-success-500 hover:text-gray-50`,
      danger: tw`border-danger-500 hover:bg-danger-500 hover:text-gray-50`,
    },

File: src/themes/theme.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+      focus: tw`focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500`,
+    },
+    color: {
+      primary: tw`bg-primary-500 hover:bg-primary-600 border-primary-500`,
+      success: tw`bg-success-500 hover:bg-success-600 border-success-500`,
+      danger: tw`bg-danger-600 hover:bg-danger-700 border-danger-600`,
+    },
+    border: {
+      gray: tw`border border-gray-300 text-gray-700 bg-white-100 hover:bg-gray-50`,
+      grey: tw`border border-gray-300 text-gray-700 bg-white-100 hover:bg-gray-50`,
```
Comment (thecodeclimber) [inline src/themes/theme.js]: Is there a way that we can make this DRYer?  There are two ways to write the word "grey" (with an "a" or with an "e").  I want to cover both possibilities.  For now, I just duplicated the code twice -- is there a better way to do this?

File: src/components/button/Button.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+const StyledButton = styled.button(
+  ({ className, styles: { block, color, border, shape, size } = {} }) => [
+    theme.button.default.base,
+    theme.button.default.color,
+    theme.button.default.hover,
+    theme.button.default.focus,
+    block && tw`w-full h-full`,
+    color && theme.button.color[color],
+    border === "gray" && theme.button.border[border],
```
Comment (thecodeclimber) [inline src/components/button/Button.js]: Border has more options then just `gray`.  There is the following:

* gray
* grey
* transparent
* primary
* success
* danger

And whatever option is chosen we need to add in `base` (see my comments for theme.js)

**Review by ghost — CHANGES_REQUESTED** [review]

File: src/themes/theme.js (hunk 4 above)
Comment (MaksymTaratuta55) [reply src/themes/theme.js]: @MaksymTaratuta55, only add it as constant:
```
const someConstant = tw`border border-gray-300 text-gray-700 bg-white-100 hover:bg-gray-50`;

{
...,
gray: someConstant,
grey: someConstant
}
```

File: src/themes/theme.js (hunk 3 above)
Comment (MaksymTaratuta55) [reply src/themes/theme.js]: fixed

File: src/components/button/Button.js (hunk 5 above)
Comment (MaksymTaratuta55) [reply src/components/button/Button.js]: fixed

File: src/themes/theme.js (hunk 4 above)
Comment (MaksymTaratuta55) [reply src/themes/theme.js]: fixed

**Comment by MaksymTaratuta55** [comment]
@thecodeclimber, updated

File: package.json [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    "@emotion/styled": "^11.0.0",
+    "@headlessui/react": "^0.2.0",
+    "autoprefixer": "^10.2.1",
+    "postcss": "^8.2.4",
+    "postcss-preset-env": "^6.7.0",
+    "react": "^17.0.1",
+    "react-dom": "^17.0.1",
+    "react-hook-form": "^6.14.2",
+    "react-icons": "^4.1.0",
+    "react-scroll": "^1.8.1",
```
Comment (thecodeclimber) [inline package.json]: Please remove `react-scroll`.  We don't need it for `ramsay-ui` (but keep it for `ramsay-site`

File: src/components/button/Button.js [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+import theme from "@/themes/theme";
+import styled from "styled-components";
+
+const StyledButton = styled.button(
+  ({ className, styles: { block, color, border, shape, size } = {} }) => [
+    theme.button.default.base,
+    theme.button.default.color,
+    theme.button.default.hover,
+    theme.button.default.focus,
+    block && tw`w-full h-full`,
```
Comment (thecodeclimber) [inline src/components/button/Button.js]: Please remove `h-full`.

File: src/components/button/Button.js [hunk 8]
```diff
@@ -0,0 +1,50 @@
+import PropTypes from "prop-types";
+import tw from "twin.macro";
+import theme from "@/themes/theme";
+import styled from "styled-components";
+
+const StyledButton = styled.button(
+  ({ className, styles: { block, color, border, shape, size } = {} }) => [
```
Comment (thecodeclimber) [inline src/components/button/Button.js:7]: Can you explain to me where `styles` comes from.  I understand that you are destructuring `block, color, border...` - but I don't understand with the `styles` object comes from.

File: src/components/button/Button.js [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+  ]
+);
+
+const Button = ({ as, children, ...props }) => {
+  return (
+    <StyledButton as={as} {...props}>
+      {children}
+    </StyledButton>
+  );
+};
```
Comment (thecodeclimber) [inline src/components/button/Button.js]: Please watch this video:  https://www.loom.com/share/a2f5f191bba94df9a4e8b94864a55dfb

Thanks.

File: src/themes/theme.js [hunk 10]
```diff
@@ -0,0 +1,41 @@
+import tw from "twin.macro";
+
+const greyAndGrayBorder = tw`border-gray-300 text-gray-700 bg-white-100 hover:bg-gray-50`;
```
Comment (thecodeclimber) [inline src/themes/theme.js]: Please rename to `greyBorder` and apply below.  Thanks.

**Review by ghost — CHANGES_REQUESTED** [review]

File: src/components/button/Button.js (hunk 8 above)
Comment (MaksymTaratuta55) [reply src/components/button/Button.js:7]: I put `block, color, border..` to one prop `styles`, so when you use `<Button>` you need put them in `styles`:
```
<Button
    className="test"
    styles={{
         border: "rounded",
         color: "primary
    }}
>Text</Button>
```
Screenshot 2021-01-19 at 17 42 42

File: src/themes/theme.js (hunk 10 above)
Comment (MaksymTaratuta55) [reply src/themes/theme.js]: fixed

File: src/components/button/Button.js (hunk 7 above)
Comment (MaksymTaratuta55) [reply src/components/button/Button.js]: fixed

File: package.json (hunk 6 above)
Comment (MaksymTaratuta55) [reply package.json]: fixed

File: src/components/button/Button.js (hunk 9 above)
Comment (MaksymTaratuta55) [reply src/components/button/Button.js]: Yes, you are right, it seems {...props} works under the hood

Comment (MaksymTaratuta55) [reply src/components/button/Button.js]: Fixed

**Comment by MaksymTaratuta55** [comment]
@thecodeclimber, updated

File: package.json [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+    "@emotion/react": "^11.1.4",
+    "@emotion/server": "^11.0.0",
+    "@emotion/styled": "^11.0.0",
+    "@headlessui/react": "^0.2.0",
+    "autoprefixer": "^10.2.1",
+    "postcss": "^8.2.4",
+    "postcss-preset-env": "^6.7.0",
+    "react": "^17.0.1",
+    "react-dom": "^17.0.1",
+    "react-hook-form": "^6.14.2",
```
Comment (thecodeclimber) [inline package.json]: Please remove `react-hook-form`.

File: src/components/button/Button.js (hunk 8 above)
Comment (thecodeclimber) [reply src/components/button/Button.js:7]: Got it.  

I personally would prefer NOT to put them all inside of `styles`.  I would rather just use each one independently -- something like this:

```
<Button color="primary" shape="rounded">Click Here</Button>
```

Is there an advantage to putting it all in `styles` -- or is it just a matter of preference?

File: src/themes/theme.js [hunk 12]
```diff
@@ -0,0 +1,41 @@
+import tw from "twin.macro";
+
+const greyBorder = tw`border-gray-300 text-gray-700 bg-white-100 hover:bg-gray-50`;
```
Comment (thecodeclimber) [inline src/themes/theme.js]: Minor change -- please make it `borderGrey`.

**Review by ghost — CHANGES_REQUESTED** [review]

File: package.json (hunk 11 above)
Comment (MaksymTaratuta55) [reply package.json]: fixed

File: src/themes/theme.js (hunk 12 above)
Comment (MaksymTaratuta55) [reply src/themes/theme.js]: fixed

**Comment by MaksymTaratuta55** [comment]
@thecodeclimber, updated

**Pull request merged**